atomica/optimization.py

```diff
@@ -1541,5 +1556,5 @@ def jacfcn(x):
         raise FailedConstraint()
 
     # Confirm constraints are all satisfied
-    assert np.all((res["x"] >= lb_scaled) & (res["x"] <= ub_scaled))
+    assert np.all((res["x"] >= lb_scaled - tolerance) & (res["x"] <= ub_scaled + tolerance)), f'FAILED as {res["x"]} not within [{lb_scaled}, {ub_scaled}]'
```

Comment:
@Rowanmh The tolerance applied to the lower bound here makes me a little uncertain since often the lower bound is 0 and this would then permit values that are slightly negative. Do you have a script that encounters the error with the tolerance? It might be better to clip the values to the lower/upper bounds if they are sufficiently close and therefore sacrifice matching the total value rather than risking changing the sign

Reply:
@RomeshA I don't have a script - I guess the change would be that if outside of the bounds but within tolerance then clip it, if outside of the bounds and outside tolerance then raise an error? Sounds appropriate to avoid anything weird happening due to a 1e-15 being somewhere instead of 0.